### PR TITLE
feat: Add S2 ActionBar component

### PR DESCRIPTION
**Pull request by devongovett**
This adds the ActionBar component to S2. It works slightly differently from S1:

* It appears on top of the collection component rather than shrinking the collection and appearing below it. This is better for performance since it doesn't require an animated relayout which is especially expensive with virtualizer. To still allow users to scroll to the bottom and see the last rows, padding is added at the end of the scrollable content to make room for the action bar.
* Instead of wrapping a collection component in an `ActionBarContainer`, there is a `renderActionBar` prop of each collection. This allows it to be more integrated, giving CardView/TableView more control over how the action bar is positioned. It also means you don't need to make selection controlled anymore, and you don't need to wire up the `selectedItemsCount` and `onClearSelection` props. The selected keys are passed into `renderActionBar` so you can conditionally adjust which actions are available, and trigger actions on those items on press.

### CODE REVIEW

File: packages/@react-spectrum/s2/stories/ActionBar.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+};
+
+export const CardExample: Story = {
+  render: args => (
+    // @ts-ignore
+    <CardViewExample
+      selectionMode="multiple"
+      renderActionBar={selectedKeys => (
+        <ActionBar {...args}>
+          <ActionButtonGroup>
```
Comment (devongovett) [inline packages/@react-spectrum/s2/stories/ActionBar.stories.tsx:59]: Do we want to accept action buttons directly as children of the ActionBar or have them pass in the group themselves? Depends if we want to allow more customization.

File: packages/@react-spectrum/s2/src/ActionBar.tsx [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+  let {isEmphasized, selectedItemCount = 0, children, onClearSelection, isExiting} = props;
+  let stringFormatter = useLocalizedStringFormatter(intlMessages, '@react-spectrum/s2');
+
+  // Store the last count greater than zero so that we can retain it while rendering the fade-out animation.
+  let [lastCount, setLastCount] = useState(selectedItemCount);
+  if ((selectedItemCount === 'all' || selectedItemCount > 0) && selectedItemCount !== lastCount) {
+    setLastCount(selectedItemCount);
+  }
+  
+  let staticColor = isEmphasized ? 'black' as const : undefined;
```
Comment (devongovett) [inline packages/@react-spectrum/s2/src/ActionBar.tsx:120]: Somehow we need to make this swap to white in dark theme and I haven't figured out how to do that yet. We don't always have a way of reading the current theme via JS anymore, it's all in CSS.

File: packages/@react-spectrum/s2/src/ActionBar.tsx [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+  }, [scrollRef]);
+
+  useResizeObserver({
+    ref: scrollRef,
+    onResize: updateScrollbarWidth
+  });
+
+  let {keyboardProps} = useKeyboard({
+    onKeyDown(e) {
+      if (e.key === 'Escape') {
```
Comment (snowystinger) [inline packages/@react-spectrum/s2/src/ActionBar.tsx:143]: continue propagation for everything else?

Comment (snowystinger) [inline packages/@react-spectrum/s2/src/ActionBar.tsx:143]: continue propagation for everything else?

File: packages/@react-spectrum/s2/src/ActionBar.tsx [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    }
+  }, [stringFormatter, scrollRef]);
+
+  return (
+    <FocusScope restoreFocus>
+      <div
+        ref={ref}
+        {...keyboardProps}
+        className={actionBarStyles({isEmphasized, isInContainer: !!scrollRef, isExiting})}
+        style={{insetInlineEnd: `calc(var(--insetEnd) + ${scrollbarWidth + 'px'})`}}>
```
Comment (snowystinger) [inline packages/@react-spectrum/s2/src/ActionBar.tsx:165]:
```suggestion
        style={{insetInlineEnd: `calc(var(--insetEnd) + ${scrollbarWidth}px)`}}>
```

File: packages/@react-spectrum/s2/src/ActionBar.tsx (hunk 3 above)
Comment (devongovett) [reply packages/@react-spectrum/s2/src/ActionBar.tsx:143]: Maybe we should have a higher level API for `useKeyboard` to make this automatic... could be based on key codes or something

```jsx
useKeyboard({
  Enter(e) {
    // Do whatever
   }
});

Comment (snowystinger) [reply packages/@react-spectrum/s2/src/ActionBar.tsx:143]: it'd be nice if it also didn't handle Shift+Enter unless explicitly declared

**Review by snowystinger — APPROVED** [review]

**Pull request merged**